### PR TITLE
Unnecessary columns are not read any more in Consensus Commit

### PR DESCRIPTION
**Pull request by brfrn169**
After this change, when reading a record in Consensus Commit, we filter out the before image columns to reduce the query cost (disk I/O and network I/O). Please take a look!

### CODE REVIEW

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java [hunk 1]
```diff
@@ -105,6 +105,12 @@ public void delete(Delete delete) {
 
   private Optional<TransactionResult> getFromStorage(Get get) throws CrudException {
     try {
+      // only get after image columns
+      get.clearProjections();
+      LinkedHashSet<String> afterImageColumnNames =
+          tableMetadataManager.getTransactionalTableMetadata(get).getAfterImageColumnNames();
+      get.withProjections(afterImageColumnNames);
+
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java:113]: Here, I added projections for after image columns to filter out the before image columns.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java [hunk 2]
```diff
@@ -114,6 +120,12 @@ public void delete(Delete delete) {
 
   private Scanner getFromStorage(Scan scan) throws CrudException {
     try {
+      // only get after image columns
+      scan.clearProjections();
+      LinkedHashSet<String> afterImageColumnNames =
+          tableMetadataManager.getTransactionalTableMetadata(scan).getAfterImageColumnNames();
+      scan.withProjections(afterImageColumnNames);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java:127]: Ditto.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java [hunk 3]
```diff
@@ -35,22 +39,61 @@ public RecoveryHandler(
...
     LOGGER.debug("recovering for {}", result.getId());
+
+    // as the result doesn't have before image columns, need to get the latest result
+    Optional<TransactionResult> latestResult;
+    try {
+      latestResult = getLatestResult(selection, result);
+    } catch (ExecutionException e) {
+      LOGGER.warn("can't get the latest result", e);
+      return;
+    }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java:50]: As commented, the `result` doesn't have before image columns due to the change, we need to re-read the record again to get before image columns.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java [hunk 4]
```diff
@@ -35,22 +39,61 @@ public RecoveryHandler(
...
+
+    if (!latestResult.isPresent()) {
+      // indicates the record is deleted in another transaction
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java:53]:
```suggestion
      // indicates the record has been deleted by another transaction
```

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java [hunk 5]
```diff
@@ -35,22 +39,61 @@ public RecoveryHandler(
...
+      return;
+    }
+
+    if (latestResult.get().isCommitted()) {
+      // indicates the record is committed in another transaction
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java:58]:
```suggestion
      // indicates the record has been committed by another transaction
```

**Review by feeblefakie — DISMISSED** [review]
LGTM! Thank you!
Left minor English suggestions.

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java [hunk 6]
```diff
@@ -114,6 +120,12 @@ public void delete(Delete delete) {
 
   private Scanner getFromStorage(Scan scan) throws CrudException {
     try {
+      // only get after image columns
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java:123]:
```suggestion
      // get only after image columns
```
Just for consistency 🙇

**Review by feeblefakie — APPROVED** [review]
Still LGTM!
Sorry, I found the same comment so I added one more suggestion.
You can merge it after it's applied. 🙇

File: core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java (hunk 6 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java:123]: Ah sorry. I missed it. Thank you!

**Pull request merged**